README.md
<img align="right" src="http://media.taskcluster.net/logo/logo-96x120.png" />

# slugid - Compressed UUIDs for Node.js

A node.js module for generating v4 UUIDs and encoding them into 22 character
URL-safe base64 slug representation (see [RFC 4648 sec.
5](http://tools.ietf.org/html/rfc4648#section-5)).

Slugs are url-safe base64 encoded v4 uuids, stripped of base64 `=` padding.
They are generated with the [uuid package](https://www.npmjs.com/package/uuid) which
is careful to use a cryptographically strong random number generator on platforms
that make one available.

There are two methods for generating slugs - `slugid.v4()` and `slugid.nice()`.

The `slugid.v4()` method returns a slug from a randomly generated v4 uuid. The
`slugid.nice()` method returns a v4 slug which conforms to a set of "nice"
properties. At the moment the only "nice" property is that the slug starts with
[A-Za-f], which in turn implies that the first (most significant) but of its
associated uuid is set to 0.

The purpose of the `slugid.nice()` method is to support having slugids which
can be used in more contexts safely. Regular slugids can safely be used in
urls, and for example in AMQP routing keys. However, slugs beginning with `-`
may cause problems when used as command line parameters.

In contrast, slugids generated by the `slugid.nice()` method can safely be used
as command line parameters. This comes at a cost to entropy (121 bits vs 122
bits for regular v4 slugs).

Slug consumers should consider carefully which of these two slug generation
methods to call. Is it more important to have maximum entropy, or to have
slugids that do not need special treatment when used as command line
parameters? This is especially important if you are providing a service which
supplies slugs to unexpecting tool developers downstream, who may not realise
the risks of using your regular v4 slugs as command line parameters, especially
since this would arise only as an intermittent issue (one time in 64).

Generated slugs take the form `[A-Za-z0-9_-]{22}`, or more precisely:

* `slugid.v4()` slugs conform to
  `[A-Za-z0-9_-]{8}[Q-T][A-Za-z0-9_-][CGKOSWaeimquy26-][A-Za-z0-9_-]{10}[AQgw]`

* `slugid.nice()` slugs conform to
  `[A-Za-f][A-Za-z0-9_-]{7}[Q-T][A-Za-z0-9_-][CGKOSWaeimquy26-][A-Za-z0-9_-]{10}[AQgw]`

RFC 4122 defines the setting of 6 bits of the v4 UUID which determines these
regular expressions.

```js
var slugid = require('slugid');

// Generate "nice" URL-safe base64 encoded UUID version 4 (random)
var slug = slugid.nice(); // a8_YezW8T7e1jLxG7evy-A
```

Encode / Decode
---------------
```js
var slugid = require('slugid');

// Generate URL-safe base64 encoded UUID version 4 (random)
var slug = slugid.v4();

// Get UUID on the form xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx
var uuid = slugid.decode(slug);

// Compress to slug again
assert(slug == slugid.encode(uuid));
```

License
-------
The `slugid` library is released on the MIT license, see the `LICENSE` for
complete license.


CHANGELOG.md
# Changelog for slugid

## v3.1.0 (2022-09-05)

* Buffer usage is [optional](https://github.com/taskcluster/slugid/pull/24) for browsers 

## v3.0.0 (2021-08-06)

* Drop support for node 4 (Node v6.17.1 is the earliest tested release)
* Add support for node 12, 14, and 16
* Remove unmaintained [uuid-parse](https://www.npmjs.com/package/uuid-parse)
* Update to [uuid](https://www.npmjs.com/package/uuid) 8.3.2, which now
  includes UUID parsing functions (again)
* Use ``Buffer.from(bytes)`` instead of the
  [insecure](https://nodejs.org/en/docs/guides/buffer-constructor-deprecation/)
  ``new Buffer(bytes)``
* Switch from TravisCI to Taskcluster for testing pull requests
* Add Code of Conduct, Contributing Guide, and Changelog

## v2.0.0 (2018-05-07)

* Drop support for node 0.10, 0.11
* Add support for node 4, 6, 8, and 10
* Update to [uuid](https://www.npmjs.com/package/uuid) 3.2.1
* Add [uuid-parse](https://www.npmjs.com/package/uuid-parse) 1.0.0 to parse
  UUIDs (instead of ``uuid``)

## v1.1.0 (2015-08-27)

* Add `slugid.nice()`, which ensures the first letter of the slug begins with a
  letter, allowing it to be used in more contexts
* Update to [uuid](https://www.npmjs.com/package/uuid) 2.0.1

## v1.0.3 (2014-10-06)

* Remove extra files not needed for the browser variant built with Browserify

## v1.0.2 (2014-08-19)

* Add a variant that runs in a browser using [Browserify](https://browserify.org/) 5.9.1

## v1.0.1 (2014-03-28)

* Add ``index.js`` with proper exports.

## v1.0.0 (2014-03-27)

* Initial release of code ported from Taskcluster



slugid_test.js
// The MIT License (MIT)
//
// Copyright (c) 2014 Jonas Finnemann Jensen
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

// Allows the tests to be run in environments with and without `Buffer`.
if (process.env.NO_BUFFER === "1") {
  delete global.Buffer;
  console.log('\x1b[33m' + 'Removed `Buffer` from globalThis.' + '\x1b[0m');
}

var slugid = require('./slugid');
var uuidv4 = require('uuid').v4;

/**
 * Test that we can correctly encode a "non-nice" uuid (with first bit set) to
 * its known slug. The specific uuid was chosen since it has a slug which
 * contains both `-` and `_` characters.
 */
exports.encodeTest = function(test) {
  test.expect(1);

  // 10000000010011110011111111001000110111111100101101001011000001101000100111111011101011101111101011010101111000011000011101010100....
  // <8 ><0 ><4 ><f ><3 ><f ><c ><8 ><d ><f ><c ><b ><4 ><b ><0 ><6 ><8 ><9 ><f ><b ><a ><e ><f ><a ><d ><5 ><e ><1 ><8 ><7 ><5 ><4 >
  // < g  >< E  >< 8  >< _  >< y  >< N  >< _  >< L  >< S  >< w  >< a  >< J  >< -  >< 6  >< 7  >< 6  >< 1  >< e  >< G  >< H  >< V  >< A  >
  var uuid_ = '804f3fc8-dfcb-4b06-89fb-aefad5e18754';
  var expectedSlug = "gE8_yN_LSwaJ-6761eGHVA";

  // Encode
  var actualSlug = slugid.encode(uuid_);

  // Test that it encoded correctly
  test.ok(expectedSlug == actualSlug, "UUID not correctly encoded into slug: '" + expectedSlug + "' != '" + actualSlug + "'");

  test.done();
};

/**
 * Test that we can decode a "non-nice" slug (first bit of uuid is set) that
 * begins with `-`
 */
exports.decodeTest = function(test) {
  test.expect(1);

  // 11111011111011111011111011111011111011111011111001000011111011111011111111111111111111111111111111111111111111111111111111111101....
  // <f ><b ><e ><f ><b ><e ><f ><b ><e ><f ><b ><e ><4 ><3 ><e ><f ><b ><f ><f ><f ><f ><f ><f ><f ><f ><f ><f ><f ><f ><f ><f ><d >
  // < -  >< -  >< -  >< -  >< -  >< -  >< -  >< -  >< Q  >< -  >< -  >< -  >< _  >< _  >< _  >< _  >< _  >< _  >< _  >< _  >< _  >< Q  >
  var slug = '--------Q--__________Q';
  var expectedUuid = "fbefbefb-efbe-43ef-bfff-fffffffffffd";

  // Decode
  var actualUuid = slugid.decode(slug);

  // Test that it is decoded correctly
  test.ok(expectedUuid == actualUuid, "Slug not correctly decoded into uuid: '" + expectedUuid + "' != '" + actualUuid + "'");

  test.done();
}

/**
 * Test that 100 v4 uuids are unchanged after encoding and then decoding them
 */
exports.uuidEncodeDecodeTest = function(test) {
  test.expect(100);

  for (i = 0; i < 100; i++) {
    // Generate uuid
    var uuid_ = uuidv4();

    // Encode
    var slug = slugid.encode(uuid_);

    // Test that decode uuid matches original
    test.ok(slugid.decode(slug) == uuid_, "Encode and decode isn't identity");
  }

  test.done();
};

/**
 * Test that 100 v4 slugs are unchanged after decoding and then encoding them.
 */
exports.slugDecodeEncodeTest = function(test) {
  test.expect(100);

  for (i = 0; i < 100; i++) {
    // Generate slug
    var slug1 = slugid.v4();

    // Decode
    var uuid_ = slugid.decode(slug1);

    // Encode
    var slug2 = slugid.encode(uuid_);

    // Test that decode uuid matches original
    test.ok(slug1 == slug2, "Decode and encode isn't identity");
  }

  test.done();
};

/**
 * Test: Make sure that all allowed characters can appear in all allowed
 * positions within the "nice" slug. In this test we generate over a thousand
 * slugids, and make sure that every possible allowed character per position
 * appears at least once in the sample of all slugids generated. We also make
 * sure that no other characters appear in positions in which they are not
 * allowed.
 *
 * base 64 encoding char -> value:
 * ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_
 * 0         1         2         3         4         5          6
 * 0123456789012345678901234567890123456789012345678901234567890123
 *
 * e.g. from this we can see 'j' represents 35 in base64
 *
 * The following comments show the 128 bits of the v4 uuid in binary, hex and
 * base 64 encodings. The 6 fixed bits (`0`/`1`) according to RFC 4122, plus
 * the first (most significant) fixed bit (`0`) are shown among the 121
 * arbitrary value bits (`.`/`x`). The `x` means the same as `.` but just
 * highlights which bits are grouped together for the respective encoding.
 *
 * schema:
 *      <..........time_low............><...time_mid...><time_hi_+_vers><clk_hi><clk_lo><.....................node.....................>
 *
 * bin: 0xxx............................................0100............10xx............................................................
 * hex:  $A <01><02><03><04><05><06><07><08><09><10><11> 4  <13><14><15> $B <17><18><19><20><21><22><23><24><25><26><27><28><29><30><31>
 *
 * => $A in {0, 1, 2, 3, 4, 5, 6, 7} (0b0xxx)
 * => $B in {8, 9, A, B} (0b10xx)
 *
 * bin: 0xxxxx..........................................0100xx......xxxx10............................................................xx0000
 * b64:   $C  < 01 >< 02 >< 03 >< 04 >< 05 >< 06 >< 07 >  $D  < 09 >  $E  < 11 >< 12 >< 13 >< 14 >< 15 >< 16 >< 17 >< 18 >< 19 >< 20 >  $F
 *
 * => $C in {A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z, a, b, c, d, e, f} (0b0xxxxx)
 * => $D in {Q, R, S, T} (0b0100xx)
 * => $E in {C, G, K, O, S, W, a, e, i, m, q, u, y, 2, 6, -} (0bxxxx10)
 * => $F in {A, Q, g, w} (0bxx0000)
 */
exports.niceSpreadTest = function(test) {
  var charsAll = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_".split('').sort().join('');
  // 0 - 31: 0b0xxxxx
  var charsC = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdef".split('').sort().join('');
  // 16, 17, 18, 19: 0b0100xx
  var charsD = "QRST".split('').sort().join('');
  // 2, 6, 10, 14, 18, 22, 26, 30, 34, 38, 42, 46, 50, 54, 58, 62: 0bxxxx10
  var charsE = "CGKOSWaeimquy26-".split('').sort().join('');
  // 0, 16, 32, 48: 0bxx0000
  var charsF = "AQgw".split('').sort().join('');
  expected = [charsC, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsD, charsAll, charsE, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsF];
  spreadTest(
    test,
    function() {
      return slugid.nice();
    },
    expected
  );
}

/**
 * This test is the same as niceSpreadTest but for slugid.v4() rather than
 * slugid.nice(). The only difference is that a v4() slug can start with any of
 * the base64 characters since the first six bits of the uuid are random.
 */
exports.v4SpreadTest = function(test) {
  var charsAll = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_".split('').sort().join('');
  // 16, 17, 18, 19: 0b0100xx
  var charsD = "QRST".split('').sort().join('');
  // 2, 6, 10, 14, 18, 22, 26, 30, 34, 38, 42, 46, 50, 54, 58, 62: 0bxxxx10
  var charsE = "CGKOSWaeimquy26-".split('').sort().join('');
  // 0, 16, 32, 48: 0bxx0000
  var charsF = "AQgw".split('').sort().join('');
  expected = [charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsD, charsAll, charsE, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsAll, charsF];
  spreadTest(
    test,
    function() {
      return slugid.v4();
    },
    expected
  );
}

/**
 * `spreadTest` runs a test against the `generator` function, to check that
 * when calling it 64*40 times, the range of characters per string position it
 * returns matches the array `expected`, where each entry in `expected` is a
 * string of all possible characters that should appear in that position in the
 * string, at least once in the sample of 64*40 responses from the `generator`
 * function */
function spreadTest(test, generator, expected) {
  // k records which characters were found at which positions. It has one entry
  // per slugid character, therefore 22 entries. Each entry is an object with
  // a property for each character found, where the value of that property is
  // the number of times that character appeared at that position in the slugid
  // in the large sample of slugids generated in this test.
  var k = [[], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], [], []];

  test.expect(1);

  // Generate a large sample of slugids, and record what characters appeared
  // where...  A monte-carlo test has demonstrated that with 64 * 20
  // iterations, no failure occurred in 1000 simulations, so 64 * 40 should be
  // suitably large to rule out false positives.
  for (i = 0; i < 64 * 40; i++) {
    var slug = generator();
    for (j = 0; j < slug.length; j++) {
      if (k[j][slug.charAt(j)] === undefined) {
        k[j][slug.charAt(j)] = 1
      } else {
        k[j][slug.charAt(j)]++;
      }
    }
  }

  // Compose results into an array `actual`, for comparison with `expected`
  var actual = [];
  for (j = 0; j < k.length; j++) {
    a = Object.keys(k[j])
    actual[j] = ""
    for (x = 0; x < a.length; x++) {
      if (k[j][a[x]] > 0) {
        actual[j] += a[x]
      }
    }
    // sort for easy comparison
    actual[j] = actual[j].split('').sort().join('');
  }

  test.ok(arraysEqual(expected, actual), "In a large sample of generated slugids, the range of characters found per character position in the sample did not match expected results.\n\nExpected: " + expected + "\n\nActual: " + actual)
  test.done();
}

/**
 * `arraysEqual` checks arrays `a` and `b` for equality
 */
function arraysEqual(a, b) {
  if (a === b) return true;
  if (a == null || b == null) return false;
  if (a.length != b.length) return false;
  for (var i = 0; i < a.length; ++i) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}


slugid.js
// The MIT License (MIT)
//
// Copyright (c) 2014 Jonas Finnemann Jensen
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var uuid = require('uuid');

/** @type {(bytes: Uint8Array) => string} */
var toBase64 = (() => {
  if (typeof Buffer !== 'undefined') {
    return (bytes) => Buffer.from(bytes).toString('base64');
  }
  return (bytes) => btoa(String.fromCharCode(...bytes));
})();

/** @type {(base64: string) => Uint8Array | Buffer} */
var fromBase64 = (() => {
  if (typeof Buffer !== 'undefined') {
    return (base64) => Buffer.from(base64, 'base64');
  }
  return (base64) => Uint8Array.from(atob(base64), c => c.charCodeAt(0));
})();

/**
 * Returns the given uuid as a 22 character slug. This can be a regular v4
 * slug or a "nice" slug.
 */
exports.encode = function(uuid_) {
  var bytes   = uuid.parse(uuid_);
  var base64 = toBase64(bytes);
  var slug = base64
              .replace(/\+/g, '-')  // Replace + with - (see RFC 4648, sec. 5)
              .replace(/\//g, '_')  // Replace / with _ (see RFC 4648, sec. 5)
              .substring(0, 22);    // Drop '==' padding
  return slug;
};

/**
 * Returns the uuid represented by the given v4 or "nice" slug
 */
exports.decode = function(slug) {
  var base64 = slug
                  .replace(/-/g, '+')
                  .replace(/_/g, '/')
                  + '==';
  return uuid.stringify(fromBase64(base64));
};

/**
 * Returns a randomly generated uuid v4 compliant slug
 */
exports.v4 = function() {
  var bytes   = uuid.v4(null, new Uint8Array(16));
  var base64 = toBase64(bytes);
  var slug = base64
              .replace(/\+/g, '-')  // Replace + with - (see RFC 4648, sec. 5)
              .replace(/\//g, '_')  // Replace / with _ (see RFC 4648, sec. 5)
              .substring(0, 22);    // Drop '==' padding
  return slug;
};

/**
 * Returns a randomly generated uuid v4 compliant slug which conforms to a set
 * of "nice" properties, at the cost of some entropy. Currently this means one
 * extra fixed bit (the first bit of the uuid is set to 0) which guarantees the
 * slug will begin with [A-Za-f]. For example such slugs don't require special
 * handling when used as command line parameters (whereas non-nice slugs may
 * start with `-` which can confuse command line tools).
 *
 * Potentially other "nice" properties may be added in future to further
 * restrict the range of potential uuids that may be generated.
 */
exports.nice = function() {
  var bytes   = uuid.v4(null, new Uint8Array(16));
  bytes[0] = bytes[0] & 0x7f;  // unset first bit to ensure [A-Za-f] first char
  var base64 = toBase64(bytes);
  var slug = base64
              .replace(/\+/g, '-')  // Replace + with - (see RFC 4648, sec. 5)
              .replace(/\//g, '_')  // Replace / with _ (see RFC 4648, sec. 5)
              .substring(0, 22);    // Drop '==' padding
  return slug;
};
